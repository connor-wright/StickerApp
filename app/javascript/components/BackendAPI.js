export function SearchImgs(query){
  return new Promise((resolve, reject) => 
  {
    fetch('/v1/imgur_api/search/' + query)
      .then(res => res.json())
      .then(result => {
        resolve(result);
      });
  });
}

export function GetImgs() {
    return new Promise((resolve, reject) => 
    {
      fetch('/v1/stickers')
        .then(res => res.json())
        .then((result) => {
          resolve(result);
        }, (error) => {
          reject(error);
        });
    });
}

export function GetImgByID(id){
  return new Promise((resolve, reject) => 
    {
      fetch(`/v1/imgur_api/?photo_id=${id}`)
        .then(res => {console.log(res.body); return res.json()})
        .then((result) => {
          resolve(result);
        }, (error) => {
          reject(error);
        });
    });
}

export function PostNewPhoto(photo){
  return new Promise((resolve, reject) => 
    {
      $.ajax({
            url: '/v1/sticker',
            type: 'post',
            dataType: 'json',
            data: {photo: photo},
            success: function(photo) {
              //add photo to stickers
              resolve(photo);
            },
            error: function(error){
              reject(error);
            }
          });
    });
}