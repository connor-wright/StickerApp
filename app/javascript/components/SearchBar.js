import React from "react";
import "./APIInterface";
import SearchPhoto from "./SearchPhoto";
import {SearchImgs} from "./APIInterface"

class SearchBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {value: '', stickers: []};
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.onClick      = this.onClick.bind(this);
  }
  
  onClick(xi) {
    const images = this.state.stickers;
    this.props.setActiveId(images[xi].id);
    this.setState({stickers: 
      images.map((image, index) => 
        ({
          active: index == xi? true: false, id: image.id, url: image.url
        })
      )
    });
  }
  
  handleChange(event) {
    this.setState({value: event.target.value});
  }
  
  handleSubmit(event) {
    if(this.state.value)
    {
      let AddStickers = (photos) => this.setState(() => ({
        stickers: photos.map(photo => 
        {
          return {active: false, id: photo.id, url: photo.src.small};
        })
      }));
      SearchImgs(this.state.value).then((result) => AddStickers(result.photos));
    }
    event.preventDefault();
  }
  
  render () {
    const {stickers} = this.state;
    return (
      <div className='SearchBar'>
        <form onSubmit={this.handleSubmit}>
          <label>
            search:
            <input type="text" data-testid='searchInput' value={this.state.value} onChange={this.handleChange} />
          </label>
            <input type="submit" value="Submit" />
          </form>
          {stickers.map((image, xi) =>
            <SearchPhoto 
              data-testid='searchPhoto'
              setActiveId={this.props.setActiveId}
              url={image.url}
              xi={xi}
              key={xi}
              active={image.active}
              onClick={this.onClick}/>
          )}
      </div>
    );
  }
}

export default SearchBar;