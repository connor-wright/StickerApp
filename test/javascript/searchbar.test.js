import React from 'react';
import SearchBar from 'SearchBar';
import {SearchImgs} from "BackendAPI";
import { mount } from 'enzyme';

jest.mock('BackendAPI');
const runAllPromises = () => new Promise(setImmediate);

let activeId;
let setActiveId;
let searchBar;

beforeEach(() =>{
  searchBar = mount(<SearchBar searchImgs={SearchImgs} />);
  activeId = '';
  setActiveId = newActiveId => activeId = newActiveId;
});

afterEach(() => {
  searchBar.unmount();
});

describe('<SearchBar />', () =>{
  it('should search triggers photo search to backend', async () =>{
    //arrange
    //act
    await searchBar.find('.searchInput').simulate('change', {target: {value: '23'}});
    await searchBar.find('.searchInput').simulate('submit');
    expect(SearchImgs).toHaveBeenCalled();
    
    //assert
    await runAllPromises();
    searchBar.update();
    expect(searchBar).toMatchSnapshot();
  });
});