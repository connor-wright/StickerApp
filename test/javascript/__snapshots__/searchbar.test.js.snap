// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should search triggers photo search to backend 1`] = `
<SearchBar
  searchImgs={
    [MockFunction] {
      "calls": Array [
        Array [
          "23",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  }
>
  <div
    className="SearchBar"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        search:
        <input
          className="searchInput"
          onChange={[Function]}
          type="text"
          value="23"
        />
      </label>
      <input
        type="submit"
        value="Submit"
      />
    </form>
    <SearchPhoto
      active={false}
      key="0"
      onClick={[Function]}
      url="https://imgur.com/gallery/3nSMulw"
      xi={0}
    >
      <img
        onClick={[Function]}
        src="https://imgur.com/gallery/3nSMulw"
        style={Object {}}
      />
    </SearchPhoto>
  </div>
</SearchBar>
`;
